odb_extractor_atoms.py

```diff
@@ -0,0 +1,627 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# Code for manipulating ODB Extractor json files and initial sequence atom definitions
+# Bryan Miller
+# 2021-11-24
+
+import os
+import json
+import gzip
+
+from openpyxl import Workbook
+from openpyxl import load_workbook
+
+
+fpuinst = {'GNIRS': 'instrument:slitWidth', 'GMOS-N': 'instrument:fpu'}
+
+
+def findatoms(sequence):
+    
+    qoffsets = []
+    atoms = []
+    natom = 0
+    nabba = 0
+    
+    nsteps = len(sequence)
+    
+    for ii, step in enumerate(sequence):
+        nextatom = False
+        keys = list(step.keys())
+
+        datalab = step['observe:dataLabel']
+        observe_class = step['observe:class']
+        exptime = float(step['observe:exposureTime'])
+        inst = step['instrument:instrument']
+    #     print(inst, fpuinst[inst])
+        fpu = step[fpuinst[inst]]
+        if 'instrument:filter' in keys:
+            filter_name = step["instrument:filter"]
+        else:
+            filter_name = 'None'
+        wavelength = float(step['instrument:observingWavelength'])
+        step_time = step['totalTime']/1000.
+        
+        if 'GMOS' in inst:
+            coadds = 1
+        else:
+            coadds = int(step['observe:coadds'])
+
+        disperser = step['instrument:disperser']
+
+        # Offsets
+        if 'telescope:p' in keys:
+            p = float(step['telescope:p'])
+        else:
+            p = 0.0
+        if 'telescope:q' in keys:
+            q = float(step['telescope:q'])
+        else:
+            q = 0.0
+        qoffsets.append(q)
+
+        # Guiding
+        guiding = 'park'
+        for key in keys:
+            if 'guideWith' in key:
+                if step[key] != 'park':
+                    guiding = step[key]
+                    break
+
+        # Any wavelength/filter_name change is a new atom
+        if ii == 0 or (ii > 0 and wavelength != float(sequence[ii-1]['instrument:observingWavelength'])):
+            nextatom = True
+            print('Atom for wavelength')
+                
+#         if observe_class == 'science':
+            
+        # AB
+        # ABBA
+        if nsteps >= 4 and nsteps - ii > 3 and nabba == 0:
+            # print(q, sequence[ii+1]['telescope:q'], sequence[ii+2]['telescope:q'], sequence[ii+3]['telescope:q'])
+            if q == float(sequence[ii+3]['telescope:q']) and q != float(sequence[ii+1]['telescope:q']) and                float(sequence[ii+1]['telescope:q']) == float(sequence[ii+2]['telescope:q']):
+                    nabba = 3
+                    nextatom = True
+                    print('Atom for ABBA')
+        else:
+            nabba -= 1
+        
+        if nextatom:
+            natom += 1
+            atoms.append({'id': natom, 'exec_time': 0.0, 'prog_time': 0.0, 'part_time': 0.0, 'filter': filter_name,
+                          'fpu': fpu, 'disperser': disperser, 'wavelength': wavelength, 'guiding': guiding})
```

Comment:
Seems to me that we should have an `Atom` dataclass that has these fields instead of a dict? And `filter`, `fpu`, and `disperser` should all be instances of `Resource` in the mini-model? Does anyone have any objections if I change this for consistency?

Reply:
Sure. I'm just using a dictionary for convenience, not to suggest that is the way to model atoms.